utils/validators.ts

```diff
@@ -0,0 +1,259 @@
+  return validateProperties(attributeModelProperties, attribute)
+}
+
+/**
+ * Checks the validity of an Category object. Does not validate child objects.
+ * @param category The Category object to test
+ * @returns A ValidationResult object
+ */
+export function validateCategory(category: any) {
+  return validateProperties(categoryModelProperties, category)
+}
+
+/**
+ * Checks the validity of an ItemDefinition object. Does not validate child objects.
+ * @param itemDefinition The ItemDefinition object to test
+ * @returns A ValidationResult object
+ */
+export function validateItemDefinition(itemDefinition: any) {
+  return validateProperties(itemDefinitionModelProperties, itemDefinition)
+}
+
+/**
+ * Checks the validity of an InventoryItem object. Does not validate child objects.
+ * @param inventoryItem The InventoryItem object to test
+ * @returns A ValidationResult object
+ */
+export function validateInventoryItem(inventoryItem: any) {
+  return validateProperties(inventoryItemModelProperties, inventoryItem)
+}
+
+/**
+ * Checks the validity of an User object. Does not validate child objects.
+ * @param user The User object to test
+ * @returns A ValidationResult object
+ */
+export function validateUser(user: any) {
+  return validateProperties(userModelProperties, user)
+}
+
+/**
+ * Validates a given id to ensure it is a valid ObjectId. Throws an error if invalid.
+ * @param id - The ObjectId to validate
+ * @returns true if the objectId was valid, false if not
+ */
+export function validateObjectId(id: string) {
+  if (!id) return false
+  if (ObjectId.isValid(id)) {
+    if (String(new ObjectId(id)) === id) return true
+  }
+  return false
+}
+
+/**
+ * Generic object validator that checks for required fields, extra fields, and type mismatches
+ * @param modelProperties An array of Property objects
+ * @param obj The object to validate
+ * @returns A ValidationResult object
+ */
+function validateProperties(
+  modelProperties: Property[],
+  obj: any
+): ValidationResult {
+  let result: ValidationResult = {
+    success: false,
+  }
+
+  if (typeof obj !== 'object') {
+    result.message = 'Validation target is not an instance of an object'
+    return result
+  }
+
+  // ensures all required properties are present
+  let isValidObject = true
+  for (const modelProp of modelProperties) {
+    if (
+      isValidObject &&
+      modelProp.required &&
+      !obj.hasOwnProperty(modelProp.key)
+    ) {
+      result.message = `Missing required property '${modelProp.key}'`
+      return result
+    }
+  }
+
+  // checks validity of all object properties.
+  for (const objKey in obj) {
+    let i = modelProperties.findIndex((prop) => prop.key === objKey)
+
+    // property is not part of server model
+    if (i < 0) {
+      console.error()
+      result.message = `Property '${objKey}' is not a valid property`
+      return result
+    }
+
+    // type mismatch
+    if (!modelProperties[i].types.includes(typeof obj[objKey])) {
```

Comment:
im not sure i understand why `includes` is used here. that is an array method right? `types` seems to always just be a string. maybe you meant to `split('|')` first?